src/libs_3rdparty/QSpec/src/primitives/GTAction.cpp

```diff
@@ -18,118 +18,62 @@
-        QAbstractButton *tb = qobject_cast<QAbstractButton *>(w);
-        if (tb) {
-            return tb;
-        }
+QAbstractButton *GTAction::button(GUITestOpStatus &os, const QString &objectName, QObject *parent, const GTGlobals::FindOptions &options) {
+    QAction *action = findAction(os, objectName, parent, options);
+    if (action == nullptr) {
+        return nullptr;
     }
-
-    return NULL;
+    QAbstractButton *resultButton = button(os, action);
+    GT_CHECK_RESULT(resultButton != nullptr || !options.failIfNotFound, "Button with object name not found: " + objectName, nullptr);
+    return resultButton;
 }
 #undef GT_METHOD_NAME
 
 #define GT_METHOD_NAME "button"
-QAbstractButton *GTAction::button(GUITestOpStatus &os, const QAction *a, QObject *parent) {
-    GT_CHECK_RESULT(a != NULL, "action is NULL", NULL);
-
-    QList<QWidget *> associated = a->associatedWidgets();
-    foreach (QWidget *w, associated) {
-        QAbstractButton *tb = qobject_cast<QAbstractButton *>(w);
-        if (tb) {
-            if (parent) {
-                QList<QToolButton *> childButtons = parent->findChildren<QToolButton *>();  // da. daa.
-                if (childButtons.contains(dynamic_cast<QToolButton *>(tb))) {
-                    return tb;
-                }
-            } else {
-                return tb;
-            }
+QAbstractButton *GTAction::button(GUITestOpStatus &os, const QAction *action) {
+    GT_CHECK_RESULT(action != nullptr, "action is NULL", nullptr);
+    QList<QWidget *> associatedWidgets = action->associatedWidgets();
+    for (QWidget *associatedWidget : associatedWidgets) {
+        if (auto button = qobject_cast<QAbstractButton *>(associatedWidget)) {
+            return button;
         }
     }
-
-    return NULL;
+    return nullptr;
 }
 #undef GT_METHOD_NAME
 
 #define GT_METHOD_NAME "findAction"
-QAction *GTAction::findAction(GUITestOpStatus &os, const QString &actionName, QObject *parent, const GTGlobals::FindOptions &options) {
-    if (parent == NULL) {  // If parent null, then searching for at QMainWindows
-        QList<QAction *> list;
-        foreach (QWidget *parent, GTMainWindow::getMainWindowsAsWidget(os)) {
-            if (parent->findChild<QAction *>(actionName) != NULL) {
-                list.append(parent->findChild<QAction *>(actionName));
-            }
-        }
-        GT_CHECK_RESULT(list.count() < 2, QString("There are %1 actions with this text").arg(list.count()), NULL);
-        if (options.failIfNotFound) {
-            GT_CHECK_RESULT(list.count() != 0, "action not found", NULL);
-            return list.takeFirst();
-        }
-        return NULL;
+QAction *GTAction::findAction(GUITestOpStatus &os, const QString &objectName, QObject *parent, const GTGlobals::FindOptions &options) {
+    QList<QAction *> actions = GTWidget::findChildren<QAction>(os, parent, [objectName](QAction *action) -> bool { return action->objectName() == objectName; });
```

Comment:
`[&objectName]`

Reply:
nit: you don't need to write a lambda return type if this type is primitive (like `bool`, `int`, etc)

Reply:
The same about the all similar cases below

Reply:
objectName -> fixed
types on lambdas -> answered in the separate thread.